### PR TITLE
Beacons request mapper

### PR DESCRIPTION
**Pull request by zackads**
## Context

PATCH requests from the `BeaconsGateway` need to satisfy the API contract defined on the [ticket](https://trello.com/c/Sw30drf0).  This PR implements a `requestMapper` that translates a domain entity (`IBeacon`) to an API request (`IBeaconRequest`).

## Changes in this pull request

- Test that BeaconsGateway calls its BeaconResponseMapper correctly (this was missed in my previous PR)
- Have BeaconsGateway call a (mocked) BeaconRequestMapper
- Implement the BeaconRequestMapper.
- Set Jest to globally clear mocks in between test, then remove ugly `beforeEach(() => jest.clearAllMocks()`.

## Guidance to review

See inline comments.

## Link to Trello card

https://trello.com/c/Sw30drf0

## Things to check

- [ ] Does this definitely match how the PATCH endpoint is implemented in the API?

### CODE REVIEW

File: src/gateways/BeaconsGateway.test.ts [hunk 1]
```diff
@@ -2,28 +2,31 @@ import axios from "axios";
...
-
 describe("BeaconsGateway", () => {
   const beaconResponseMapper: IBeaconResponseMapper = {
     map: jest.fn(),
   };
 
-  describe("getAllBeacons()", () => {
-    beforeEach(() => {
-      jest.clearAllMocks();
-    });
```
Comment (zackads) [inline src/gateways/BeaconsGateway.test.ts:23]: See https://github.com/facebook/jest/issues/4000, https://jestjs.io/docs/configuration#clearmocks-boolean and https://jestjs.io/docs/jest-object#jestclearallmocks.

TL;DR - I've set Jest to clear all mocks in between tests by default and removed all `jest.clearAllMocks()` calls from our test suite to improve readability and eliminate mock state bleeding between tests.

File: src/gateways/mappers/BeaconRequestMapper.test.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+          model: "Beaconater 3000",
+          manufacturerSerialNumber: "BCNTR3000",
+          chkCode: "45YU",
+          batteryExpiryDate: "2021-04-20",
+          lastServicedDate: "20-04-2021",
+        },
+      },
+    });
+  });
+});
```
Comment (zackads) [inline src/gateways/mappers/BeaconRequestMapper.test.ts:106]: `id` exists on the `IBeacon` coming in, but I've ignored it as an attribute on the `BeaconRequest` going out (it's still there under `data.id`.  Is this the right approach?  Or should `id` appear again in `attributes`?

File: src/gateways/mappers/BeaconRequestMapper.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+    return {
+      data: {
+        type: "beacon",
+        id: beaconId,
+        attributes: attributes,
+      },
+    };
+  }
+}
```
Comment (zackads) [inline src/gateways/mappers/BeaconRequestMapper.ts:36]: There's an even uglier ES6 syntax for achieving this [here](https://medium.com/@mikeh91/conditionally-adding-keys-to-javascript-objects-using-spread-operators-and-short-circuit-evaluation-acf157488ede).
```typescript
const buildAnObjectFromAQuery = query => ({
  ...query.foo && { foo: query.foo },
  ...query.bar && { bar: query.bar },
});
```

I chose to go with simple inline if statements for readability and to make the mapping explicit.

File: src/gateways/mappers/IBeaconRequest.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+      createdDate?: string;
+      model?: string;
+      manufacturerSerialNumber?: string;
+      chkCode?: string;
+      protocolCode?: string;
+      batteryExpiryDate?: string;
+      lastServicedDate?: string;
+    };
+  };
+}
```
Comment (zackads) [inline src/gateways/mappers/IBeaconRequest.ts:19]: `relationships` key not included, waiting until we get to updating relationships between a beacon and its owner and uses etc. before implementing.

File: src/gateways/mappers/BeaconRequestMapper.test.ts (hunk 2 above)
Comment (Ninamma) [reply src/gateways/mappers/BeaconRequestMapper.test.ts:106]: I think that makes sense, I'd say the backend/service thinks that a Beacon looks like this: https://github.com/mcagov/beacons-service/blob/main/src/test/resources/fixtures/getBeaconResponse.json - so I think just `data.id` is good.

File: src/gateways/mappers/BeaconRequestMapper.ts (hunk 3 above)
Comment (stuart-madetech) [reply src/gateways/mappers/BeaconRequestMapper.ts:36]: Fine :)

File: src/gateways/BeaconsGateway.test.ts (hunk 1 above)
Comment (stuart-madetech) [reply src/gateways/BeaconsGateway.test.ts:23]: Good call. Reset state

**Review by stuart-madetech — CHANGES_REQUESTED** [review]
Lets change the request patch to only send actually updated fields rather than all :)

File: src/panels/beaconSummaryPanel/BeaconSummaryPanel.tsx [hunk 5]
```diff
@@ -42,9 +43,12 @@ export const BeaconSummaryPanel: FunctionComponent<IBeaconSummaryProps> = ({
...
   }, [userState, beaconId, beaconsGateway]);
 
-  const handleSave = async (beacon: IBeacon): Promise<void> => {
+  const handleSave = async (updatedBeacon: Partial<IBeacon>): Promise<void> => {
     try {
-      await beaconsGateway.updateBeacon(beacon.id, beacon);
+      await beaconsGateway.updateBeacon(
+        beacon.id,
+        diffObjValues(beacon, updatedBeacon)
+      );
```
Comment (zackads) [inline src/panels/beaconSummaryPanel/BeaconSummaryPanel.tsx:51]: I chose to place the responsibility for only submitting updated values in the `Panel` because:
- The `Gateway` should only be responsible for making calls, not manipulating data;
- The `Mapper` should only be responsible for translating between data formats, not making calls;

This left me with either the `EditingState` or the `Panel` components.
- Putting the `diffObjValues` in the `EditingState` felt iffy for me because it interferes with the basic pattern of all form values being submitted when the form is submitted.
- The `Panel` has access to the original beacon, the updated beacon and the gateway.  It felt natural for the `Panel` to receive the form submission, filter out unchanged fields and then call the relevant method on the gateway.

Grateful for thoughts!  Is there another design I haven't thought of?

**Review by stuart-madetech — APPROVED** [review]
